File: product-service/src/main/java/com/intexsoft/webshop/productservice/controller/VendorController.java
package com.intexsoft.webshop.productservice.controller;

import com.intexsoft.webshop.productservice.dto.vendor.VendorCreateDto;
import com.intexsoft.webshop.productservice.dto.vendor.VendorUpdateDto;
import com.intexsoft.webshop.productservice.dto.vendor.VendorDto;
import com.intexsoft.webshop.productservice.service.VendorService;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Positive;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

import static com.intexsoft.webshop.productservice.util.JsonUtils.getAsString;
import static java.lang.Integer.MAX_VALUE;

@RestController
@RequestMapping("/vendor")
@RequiredArgsConstructor
@Slf4j
@Validated
public class VendorController {
    private final VendorService vendorService;

    @PostMapping
    public ResponseEntity<VendorDto> createVendor(@RequestBody @Valid VendorCreateDto vendorCreateDto) {
        log.info("IN: request to create a new vendor received. Request body = {}", getAsString(vendorCreateDto));
        VendorDto createdVendor = vendorService.createVendor(vendorCreateDto);
        log.info("OUT: new vendor created successfully. Response body = {}", getAsString(createdVendor));
        return ResponseEntity.ok(createdVendor);
    }

    @GetMapping
    public ResponseEntity<List<VendorDto>> findVendors(@PageableDefault(size = MAX_VALUE) Pageable pageable) {
        log.info("IN: request to find vendors received. Page size = {}, page number = {}",
                pageable.getPageSize(), pageable.getPageNumber());
        List<VendorDto> vendorDtos = vendorService.findVendors(pageable);
        log.info("OUT: {} vendors found", vendorDtos.size());
        return ResponseEntity.ok(vendorDtos);
    }

    @GetMapping("/{vendorId}")
    public ResponseEntity<VendorDto> findVendorById(@PathVariable @Positive Long vendorId) {
        log.info("IN: request to find a vendor by id = {} received", vendorId);
        VendorDto vendorDto = vendorService.findVendorById(vendorId);
        log.info("OUT: the vendor with id = {} found successfully. Response body = {}",
                vendorId, getAsString(vendorDto));
        return ResponseEntity.ok(vendorDto);
    }

    @PutMapping("/{vendorId}")
    public ResponseEntity<VendorDto> updateVendor(@PathVariable @Positive Long vendorId,
                                                  @RequestBody @Valid VendorUpdateDto vendorUpdateDto) {
        log.info("IN: request to update a vendor with id = {} received. Request body = {}",
                vendorId, getAsString(vendorUpdateDto));
        VendorDto updatedVendorDto = vendorService.updateVendor(vendorId, vendorUpdateDto);
        log.info("OUT: the vendor with id = {} updated successfully. Response body = {}",
                vendorId, getAsString(updatedVendorDto));
        return ResponseEntity.ok(updatedVendorDto);
    }

    @DeleteMapping("/{vendorId}")
    public ResponseEntity<Void> deleteProductById(@PathVariable @Positive Long vendorId) {
        log.info("IN: request to delete a category by id = {} received", vendorId);
        vendorService.deleteVendorById(vendorId);
        log.info("OUT: the vendor with id = {} deleted successfully", vendorId);
        return ResponseEntity.noContent().build();
    }
}

File: product-service-kt/src/main/kotlin/com/intexsoft/webshop/productservicekt/controller/VendorController.kt
package com.intexsoft.webshop.productservicekt.controller

import com.intexsoft.webshop.productservicekt.dto.vendor.VendorCreateDto
import com.intexsoft.webshop.productservicekt.dto.vendor.VendorDto
import com.intexsoft.webshop.productservicekt.dto.vendor.VendorUpdateDto
import com.intexsoft.webshop.productservicekt.log
import com.intexsoft.webshop.productservicekt.mapper.VendorMapper
import com.intexsoft.webshop.productservicekt.service.VendorService
import jakarta.validation.Valid
import jakarta.validation.constraints.Positive
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/vendor")
@Validated
class VendorController(
    private val vendorService: VendorService,
    private val vendorMapper: VendorMapper
) {
    @PostMapping
    fun createVendor(@RequestBody vendorCreateDto: @Valid VendorCreateDto): ResponseEntity<VendorDto> {
        log.info("IN: request to create a new vendor received. Request body = $vendorCreateDto")
        val createdVendor: VendorDto = vendorService.createVendor(vendorCreateDto)
        log.info("OUT: new vendor created successfully. Response body = $createdVendor")
        return ResponseEntity.ok(createdVendor)
    }

    @GetMapping
    fun findVendors(@PageableDefault(size = Int.MAX_VALUE) pageable: Pageable): ResponseEntity<List<VendorDto>> {
        log.info("IN: request to find vendors received. Page size = ${pageable.pageSize}, page number = ${pageable.pageNumber}")
        val vendorDtos: List<VendorDto> = vendorService.findVendors(pageable)
        log.info("OUT: ${vendorDtos.size} vendors found")
        return ResponseEntity.ok(vendorDtos)
    }

    @GetMapping("/{vendorId}")
    fun findVendorById(@PathVariable vendorId: @Positive Long): ResponseEntity<VendorDto> {
        log.info("IN: request to find a vendor by id = $vendorId received")
        val vendorDto: VendorDto = vendorService.findVendorById(vendorId)
        log.info("OUT: the vendor with id = $vendorId found successfully. Response body = $vendorDto")
        return ResponseEntity.ok(vendorDto)
    }

    @PutMapping("/{vendorId}")
    fun updateVendor(
        @PathVariable vendorId: @Positive Long,
        @RequestBody vendorUpdateDto: @Valid VendorUpdateDto
    ): ResponseEntity<VendorDto> {
        log.info("IN: request to update a vendor with id = $vendorId received. Request body = $vendorUpdateDto")
        val updatedVendorDto: VendorDto = vendorService.updateVendor(vendorId, vendorUpdateDto)
        log.info("OUT: the vendor with id = $vendorId updated successfully. Response body = $updatedVendorDto")
        return ResponseEntity.ok(updatedVendorDto)
    }

    @DeleteMapping("/{vendorId}")
    fun deleteProductById(@PathVariable vendorId: @Positive Long): ResponseEntity<Unit> {
        log.info("IN: request to delete a category by id = $vendorId received")
        vendorService.deleteVendorById(vendorId)
        log.info("OUT: the vendor with id = $vendorId deleted successfully")
        return ResponseEntity.noContent().build()
    }
}